docs/architecture/ADR-003-Non-interactive-defaults.md

```diff
@@ -0,0 +1,691 @@
+
+We are already wrapping/unwrapping malleated transactions, so including the `share_index` as metadata using the current code is essentially as simple as adding the `share_index` to the struct. Transactions are unwrapped selectively by using a [`MalleatedTxDecoder`](https://github.com/celestiaorg/celestia-app/blob/5ac236fb1dab6628e98a505269f295c18e150b27/app/encoding/malleated_tx_decoder.go#L8-L15) or by using the [`UnwrapMalleatedTx`](https://github.com/celestiaorg/celestia-core/blob/212901fcfc0f5a095683b1836ea9e890cc952dc7/types/tx.go#L214-L237) function.
+
+```proto
+message MalleatedTx {
+bytes original_tx_hash = 1;
+bytes tx               = 2;
+uint32 share_index     = 3;
+}
+```
+
+### Refactor Share Splitting and Merging
+
+Our encoding scheme now has to actually use the metadata added to wrapped transactions.
+
+Note: In order to properly test the new encoding scheme, we have to perform identical if not very similar application logic to that in `PrepareProposal`, for this reason, we initially wanted to move the share encoding/decoding logic to the app instead of core. There were some other quality of life improvements that were also added during this refactor that are technically unrelated to these changes.
+
+We currently utilize a struct to store the state needed for lazily writing message shares. Here we add a method to it that allows for us to write namespaced padded shares.
+
+```go
+// MessageShareSplitter lazily splits messages into shares that will eventually be
+// included in a data square. It also has methods to help progressively count
+// how many shares the messages written take up.
+type MessageShareSplitter struct {
+  shares [][]NamespacedShare
+  count  int
+}
+...
+// WriteNamespacedPaddedShares adds empty shares using the namespace of the last written share.
+// This is useful to follow the message layout rules. It assumes that at least
+// one share has already been written, if not it panics.
+func (msw *MessageShareSplitter) WriteNamespacedPaddedShares(count int) {
+  if len(msw.shares) == 0 {
+      panic("Cannot write empty namespaced shares on an empty MessageShareSplitter")
+  }
+  if count == 0 {
+      return
+  }
+  lastMessage := msw.shares[len(msw.shares)-1]
+  msw.shares = append(msw.shares, namespacedPaddedShares(lastMessage[0].ID, count))
+  msw.count += count
+}
+```
+
+Now we simply combine this new functionality with the `share_index`s described above, and we can properly split and pad messages when needed. Note, the below implementation allows for `nil` to be passed as indexes. This is important, as it allows the same implementation to be used in the cases where we don't want to split messages using wrapped transactions, such as supporting older networks or when users create commitments to sign over for `MsgWirePayForData`
+
+```go
+func SplitMessages(cursor int, indexes []uint32, msgs []coretypes.Message, useShareIndexes bool) ([][]byte, error) {
+  if len(indexes) != len(msgs) {
```

Comment:
[Non Blocking][Question]
According to this:
> Note, the below implementation allows for `nil` to be passed as indexes.

If `indexes` is `nil` or `0`, it will error.  Is this intended?
```suggestion
  if len(indexes) != 0 && len(indexes) != len(msgs) {
```